diff --git a/src/menus/options-video-advanced.c b/src/menus/options-video-advanced.c
index c2e40ae04..f785e6637 100644
--- a/src/menus/options-video-advanced.c
+++ b/src/menus/options-video-advanced.c
@@ -88,6 +88,9 @@ menuitem_t OPTIONS_VideoAdvanced[] =
 	{IT_STRING | IT_CVAR, "Anisotropic", "Lower values will improve performance at a minor quality loss.",
 		NULL, {.cvar = &cv_glanisotropicmode}, 0, 0},
 
+	{IT_STRING | IT_CVAR, "Texture Supersampling", "Extra texture samples per pixel. Stops distant pixel art from crawling.",
+		NULL, {.cvar = &cv_glsupersample}, 0, 0},
+
 	{IT_SPACE | IT_NOTHING, NULL,  NULL,
 		NULL, {NULL}, 0, 0},
 
