diff --git a/src/rhi/shader_load_context.cpp b/src/rhi/shader_load_context.cpp
index ad3ffcf81..1573ac3fd 100644
--- a/src/rhi/shader_load_context.cpp
+++ b/src/rhi/shader_load_context.cpp
@@ -10,10 +10,15 @@
 
 #include "shader_load_context.hpp"
 
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
 #include <fmt/core.h>
 
 #include "../core/string.h"
 #include "../core/vector.hpp"
+#include "../w_wad.h"
 
 using namespace srb2;
 using namespace rhi;
@@ -53,3 +58,61 @@ srb2::Vector<const char*> ShaderLoadContext::get_sources_array()
 
 	return ret;
 }
+
+srb2::Vector<srb2::String> srb2::rhi::read_glsllist_sources(const char* lumpname)
+{
+	size_t buffer_size;
+	if (!W_ReadShader(lumpname, &buffer_size, nullptr))
+	{
+		throw std::runtime_error(fmt::format("Unable to find glsllist lump {}", lumpname));
+	}
+	srb2::String glsllist_lump_data;
+	glsllist_lump_data.resize(buffer_size);
+	if (!W_ReadShader(lumpname, &buffer_size, glsllist_lump_data.data()))
+	{
+		throw std::runtime_error(fmt::format("Unable to read glsllist lump {}", lumpname));
+	}
+
+	std::istringstream glsllist(glsllist_lump_data);
+	srb2::Vector<srb2::String> sources;
+	for (std::string line; std::getline(glsllist, line); )
+	{
+		if (line.empty())
+		{
+			continue;
+		}
+
+		if (line[0] == '#')
+		{
+			continue;
+		}
+
+		if (line.back() == '\r')
+		{
+			line.pop_back();
+		}
+
+		// Compat: entries not ending in .glsl should append, for new shader file lookup system
+		size_t glsl_pos = line.find(".glsl");
+		if (line.size() < 5 || glsl_pos == line.npos || glsl_pos != line.size() - 5)
+		{
+			line.append(".glsl");
+		}
+
+		size_t source_lump_size;
+		if (!W_ReadShader(line.c_str(), &source_lump_size, nullptr))
+		{
+			throw std::runtime_error(fmt::format("Unable to find glsl source lump lump {}", line));
+		}
+		srb2::String source_lump;
+		source_lump.resize(source_lump_size);
+		if (!W_ReadShader(line.c_str(), &source_lump_size, source_lump.data()))
+		{
+			throw std::runtime_error(fmt::format("Unable to read glsl source lump lump {}", line));
+		}
+
+		sources.emplace_back(source_lump);
+	}
+
+	return sources;
+}
