diff --git a/src/blua/liolib.c b/src/blua/liolib.c
index 6b44e2eb8..427345e91 100644
--- a/src/blua/liolib.c
+++ b/src/blua/liolib.c
@@ -189,7 +189,7 @@ static int CheckFileName(lua_State *L, const char *filename)
 	boolean pass = false;
 	size_t i;
 
-	if (strchr(filename, '\\'))
+	if (luaL_strchr(filename, '\\'))
 	{
 		luaL_error(L, "access denied to %s: \\ is not allowed, use / instead", filename);
 		return pushresult(L,0,filename);
@@ -202,7 +202,7 @@ static int CheckFileName(lua_State *L, const char *filename)
 			break;
 		}
 	if (strstr(filename, "./")
-		|| strstr(filename, "..") || strchr(filename, ':')
+		|| strstr(filename, "..") || luaL_strchr(filename, ':')
 		|| filename[0] == '/'
 		|| !pass)
 	{
@@ -224,7 +224,7 @@ static int io_open (lua_State *L) {
 
 	luaL_checktype(L, 3, LUA_TFUNCTION);
 
-	if (!(strchr(mode, 'r') || strchr(mode, '+')))
+	if (!(luaL_strchr(mode, 'r') || luaL_strchr(mode, '+')))
 		luaL_error(L, "open() is only for reading, use openlocal() for writing");
 
 	AddLuaFileTransfer(filename, mode);
