diff --git a/src/command.h b/src/command.h
index 70342a785..5ea58c8e3 100644
--- a/src/command.h
+++ b/src/command.h
@@ -125,6 +125,7 @@ typedef enum
                    // used on menus
 	CV_CHEAT = 2048, // Don't let this be used in multiplayer unless cheats are on.
 	CV_ALLOWLUA = 4096,/* Let this be called from Lua */
+	CV_SLIDER_SAFE = 8192, // This CVar is safe to use with a slider being changed by a touch screen.
 } cvflags_t;
 
 typedef struct CV_PossibleValue_s
@@ -218,6 +219,9 @@ void CV_StealthSet(consvar_t *var, const char *value);
 // it a setvalue but with a modulo at the maximum
 void CV_AddValue(consvar_t *var, INT32 increment);
 
+// Returns the longest PossibleValue string for this CVar
+const char *CV_LongestPossibleValue(consvar_t *var);
+
 // write all CV_SAVE variables to config file
 void CV_SaveVariables(FILE *f);
 
