diff --git a/src/console.c b/src/console.c
index 50ecfec9c..0bedc1530 100644
--- a/src/console.c
+++ b/src/console.c
@@ -109,6 +109,10 @@ static void CONS_height_Change(void);
 static void CONS_hudlines_Change(void);
 static void CONS_backcolor_Change(void);
 
+#ifdef VIRTUAL_KEYBOARD
+static void CON_ScreenKeyboardInput(char *text, size_t length);
+#endif
+
 //======================================================================
 //                   CONSOLE VARS AND COMMANDS
 //======================================================================
@@ -485,6 +489,10 @@ void CON_Init(void)
 	con_destlines = vid.height;
 	con_curlines = vid.height;
 
+#ifdef TOUCHINPUTS
+	cons_height.defaultvalue = "40";
+#endif
+
 	Unlock_state();
 
 	if (!dedicated)
@@ -728,7 +736,7 @@ void CON_ClearHUD(void)
 }
 
 // Force console to move out immediately
-// note: con_ticker will set consoleready false
+// Now sets consoleready to false
 void CON_ToggleOff(void)
 {
 	Lock_state();
@@ -744,8 +752,15 @@ void CON_ToggleOff(void)
 	CON_ClearHUD();
 	con_forcepic = 0;
 
+	consoleready = false;
+
 	I_UpdateMouseGrab();
 
+#ifdef VIRTUAL_KEYBOARD
+	if (I_KeyboardOnScreen())
+		I_CloseScreenKeyboard();
+#endif
+
 	Unlock_state();
 }
 
@@ -760,6 +775,16 @@ boolean CON_Ready(void)
 	return ready;
 }
 
+boolean CON_Allowed(void)
+{
+	return (!modeattacking && !metalrecording && !marathonmode && !chat_on);
+}
+
+void CON_Toggle(void)
+{
+	consoletoggle = true;
+}
+
 // Console ticker: handles console move in/out, cursor blinking
 //
 void CON_Ticker(void)
@@ -775,6 +800,12 @@ void CON_Ticker(void)
 	con_tick++;
 	con_tick &= 7;
 
+#ifdef VIRTUAL_KEYBOARD
+	// Close the console if the screen keyboard is not visible
+	if (con_destlines > 0 && !I_KeyboardOnScreen())
+		consoletoggle = true;
+#endif
+
 	// console key was pushed
 	if (consoletoggle)
 	{
@@ -786,14 +817,31 @@ void CON_Ticker(void)
 			con_destlines = 0;
 			CON_ClearHUD();
 			I_UpdateMouseGrab();
+#ifdef VIRTUAL_KEYBOARD
+			if (I_KeyboardOnScreen())
+				I_CloseScreenKeyboard();
+#endif
 		}
 		else
 			CON_ChangeHeight();
+
+#ifdef VIRTUAL_KEYBOARD
+		con_scrollup = 0;
+#endif
 	}
 
 	// check if console ready for prompt
 	if (con_destlines >= minheight)
+	{
+#ifdef VIRTUAL_KEYBOARD
+		if (!(consoleready && I_KeyboardOnScreen())) // Show the virtual keyboard
+		{
+			I_ShowVirtualKeyboard(NULL, 0);
+			I_SetVirtualKeyboardCallback(CON_ScreenKeyboardInput);
+		}
+#endif
 		consoleready = true;
+	}
 	else
 		consoleready = false;
 
@@ -914,6 +962,14 @@ static void CON_InputDelChar(void)
 	Unlock_state();
 }
 
+#ifdef VIRTUAL_KEYBOARD
+static void CON_ScreenKeyboardInput(char *text, size_t length)
+{
+	(void)length;
+	CON_InputAddString(text);
+}
+#endif
+
 //
 // ----
 //
@@ -1129,7 +1185,7 @@ boolean CON_Responder(event_t *ev)
 
 		// ...why shouldn't it eat the key? if it doesn't, it just means you
 		// can control Sonic from the console, which is silly
-		return true;//return false;
+		return true;
 	}
 
 	// command completion forward (tab) and backward (shift-tab)
@@ -1242,7 +1298,13 @@ boolean CON_Responder(event_t *ev)
 	if (key == KEY_ENTER)
 	{
 		if (!input_len)
+		{
+#ifdef VIRTUAL_KEYBOARD
+			if (I_KeyboardOnScreen())
+				CON_Toggle();
+#endif
 			return true;
+		}
 
 		// push the command
 		COM_BufAddText(inputlines[inputline]);
@@ -1291,6 +1353,12 @@ boolean CON_Responder(event_t *ev)
 		return true;
 	}
 
+#ifdef VIRTUAL_KEYBOARD
+	// Inputs handled elsewhere
+	if (I_KeyboardOnScreen())
+		return true;
+#endif
+
 	// allow people to use keypad in console (good for typing IP addresses) - Calum
 	if (key >= KEY_KEYPAD7 && key <= KEY_KPADDEL)
 	{
@@ -1471,7 +1539,7 @@ void CONS_Printf(const char *fmt, ...)
 		txt = malloc(8192);
 
 	va_start(argptr, fmt);
-	vsprintf(txt, fmt, argptr);
+	M_vsnprintf(txt, 8192, fmt, argptr);
 	va_end(argptr);
 
 	// echo console prints to log file
@@ -1491,12 +1559,14 @@ void CONS_Printf(const char *fmt, ...)
 
 	Unlock_state();
 
+#if !defined(__ANDROID__)
 	// if not in display loop, force screen update
-	if (refresh)
+	if (refresh && !I_AppOnBackground())
 	{
 		CON_Drawer(); // here we display the console text
 		I_FinishUpdate(); // page flip or blit buffer
 	}
+#endif
 }
 
 void CONS_Alert(alerttype_t level, const char *fmt, ...)
@@ -1508,7 +1578,7 @@ void CONS_Alert(alerttype_t level, const char *fmt, ...)
 		txt = malloc(8192);
 
 	va_start(argptr, fmt);
-	vsprintf(txt, fmt, argptr);
+	M_vsnprintf(txt, 8192, fmt, argptr);
 	va_end(argptr);
 
 	switch (level)
@@ -1544,7 +1614,7 @@ void CONS_Debug(INT32 debugflags, const char *fmt, ...)
 		txt = malloc(8192);
 
 	va_start(argptr, fmt);
-	vsprintf(txt, fmt, argptr);
+	M_vsnprintf(txt, 8192, fmt, argptr);
 	va_end(argptr);
 
 	// Again I am lazy, oh well
@@ -1724,7 +1794,7 @@ static void CON_DrawBackpic(void)
 {
 	patch_t *con_backpic;
 	lumpnum_t piclump;
-	int x, w, h;
+	INT32 x, y = 0, w, h = 0;
 
 	// Get the lumpnum for CONSBACK, STARTUP (Only during game startup) or fallback into MISSING.
 	if (con_startup)
@@ -1745,7 +1815,7 @@ static void CON_DrawBackpic(void)
 
 	// If the patch doesn't fill the entire screen,
 	// then fill the sides with a solid color.
-	if (x > 0)
+	if (x > 0 || con_startup)
 	{
 		column_t *column = &con_backpic->columns[0];
 		if (column->num_posts && !column->posts[0].topdelta)
@@ -1760,11 +1830,16 @@ static void CON_DrawBackpic(void)
 	}
 
 	// Draw the patch.
-	V_DrawCroppedPatch(x << FRACBITS, 0, FRACUNIT, FRACUNIT, V_NOSCALESTART, con_backpic, NULL,
+	if (con_startup)
+	{
+		V_DrawCroppedPatch(x << FRACBITS, y << FRACBITS, FRACUNIT, FRACUNIT, V_NOSCALESTART, con_backpic, NULL,
+			0, 0, BASEVIDWIDTH << FRACBITS, BASEVIDHEIGHT << FRACBITS);
+	}
+	else
+	{
+		V_DrawCroppedPatch(x << FRACBITS, 0, FRACUNIT, FRACUNIT, V_NOSCALESTART, con_backpic, NULL,
 			0, (BASEVIDHEIGHT - h) << FRACBITS, BASEVIDWIDTH << FRACBITS, h << FRACBITS);
-
-	// Unlock the cached patch.
-	W_UnlockCachedPatch(con_backpic);
+	}
 }
 
 // draw the console background, text, and prompt if enough place
