diff --git a/src/deh_soc.c b/src/deh_soc.c
index 343beb301..d5491369d 100644
--- a/src/deh_soc.c
+++ b/src/deh_soc.c
@@ -32,6 +32,7 @@
 #include "r_picformats.h"
 #include "r_things.h" // R_Char2Frame
 #include "r_sky.h"
+#include "s_sound.h"
 #include "fastcmp.h"
 #include "lua_script.h" // Reluctantly included for LUA_EvalMath
 #include "netcode/d_clisrv.h"
@@ -3907,6 +3908,7 @@ void readmaincfg(MYFILE *f)
 			else if (fastcmp(word, "GAMEDATA"))
 			{
 				size_t filenamelen;
+				char savegame[SAVEGAMENAMELEN];
 
 				// Check the data filename so that mods
 				// can't write arbitrary files.
@@ -3927,13 +3929,13 @@ void readmaincfg(MYFILE *f)
 				strncpy(timeattackfolder, gamedatafilename, filenamelen);
 				timeattackfolder[min(filenamelen, sizeof (timeattackfolder) - 1)] = '\0';
 
-				strcpy(savegamename, timeattackfolder);
-				strlcat(savegamename, "%u.ssg", sizeof(savegamename));
-				// can't use sprintf since there is %u in savegamename
-				strcatbf(savegamename, srb2home, PATHSEP);
+				strlcpy(savegame, timeattackfolder, sizeof(savegame));
+				strlcat(savegame, "%u.ssg", sizeof(savegame));
 
-				strcpy(liveeventbackup, va("live%s.bkp", timeattackfolder));
-				strcatbf(liveeventbackup, srb2home, PATHSEP);
+				D_DefaultSaveGameName(savegame);
+				D_DefaultLiveEventName(va("live%s.bkp", timeattackfolder)); // intentionally not ending with .ssg
+
+				D_MakeSaveGamePaths(srb2home);
 
 				gamedataadded = true;
 				titlechanged = true;
