diff --git a/src/m_anigif.c b/src/m_anigif.c
index 5ebc4892c..cb7054fd8 100644
--- a/src/m_anigif.c
+++ b/src/m_anigif.c
@@ -52,6 +52,14 @@ static boolean gif_colorprofile = false;
 static RGBA_t *gif_headerpalette = NULL;
 static RGBA_t *gif_framepalette = NULL;
 
+typedef struct
+{
+	void *pixels;
+	size_t size;
+	boolean owns_pixels;
+} gif_screen_t;
+static gif_screen_t gif_screens[2];
+
 static FILE *gif_out = NULL;
 static INT32 gif_frames = 0;
 static precise_t gif_prevframetime = 0;
@@ -513,7 +521,7 @@ static void GIF_rgbconvert(UINT8 *linear, UINT8 *scr)
 {
 	UINT8 r, g, b;
 	size_t src = 0, dest = 0;
-	size_t size = (vid.width * vid.height * 3);
+	size_t size = (vid.width * vid.height * SCREENSHOT_BITS);
 
 	InitColorLUT(&gif_colorlookup, (gif_localcolortable) ? gif_framepalette : gif_headerpalette, true);
 
@@ -523,7 +531,7 @@ static void GIF_rgbconvert(UINT8 *linear, UINT8 *scr)
 		g = (UINT8)linear[src + 1];
 		b = (UINT8)linear[src + 2];
 		scr[dest] = GetColorLUTDirect(&gif_colorlookup, r, g, b);
-		src += (3 * scrbuf_downscaleamt);
+		src += (SCREENSHOT_BITS * scrbuf_downscaleamt);
 		dest += scrbuf_downscaleamt;
 	}
 }
@@ -536,7 +544,8 @@ static void GIF_rgbconvert(UINT8 *linear, UINT8 *scr)
 static void GIF_framewrite(void)
 {
 	UINT8 *p;
-	UINT8 *movie_screen = screens[2];
+	UINT8 *base_screen = gif_screens[0].pixels;
+	UINT8 *movie_screen = gif_screens[1].pixels;
 	INT32 blitx, blity, blitw, blith;
 	boolean palchanged;
 
@@ -561,7 +570,7 @@ static void GIF_framewrite(void)
 	if (gif_optimize && gif_frames > 0 && (!palchanged))
 	{
 		// before blit movie_screen points to last frame, cur_screen points to this frame
-		UINT8 *cur_screen = screens[0];
+		UINT8 *cur_screen = base_screen;
 		GIF_optimizeregion(cur_screen, movie_screen, &blitx, &blity, &blitw, &blith);
 
 		// blit to temp screen
@@ -587,7 +596,7 @@ static void GIF_framewrite(void)
 		if (rendermode == render_opengl)
 		{
 			UINT8 *linear = HWR_GetScreenshot();
-			GIF_rgbconvert(linear, screens[0]);
+			GIF_rgbconvert(linear, base_screen);
 			free(linear);
 		}
 #endif
@@ -597,7 +606,7 @@ static void GIF_framewrite(void)
 		if (gif_frames == 0 && rendermode == render_soft)
 			I_ReadScreen(movie_screen);
 
-		movie_screen = screens[0];
+		movie_screen = base_screen;
 	}
 
 	// screen regions are handled in GIF_lzw
@@ -742,16 +751,70 @@ INT32 GIF_open(const char *filename)
 	gif_frames = 0;
 	gif_prevframetime = I_GetPreciseTime();
 	gif_delayus = 0;
+
 	return 1;
 }
 
+static void GIF_checkscreens(void)
+{
+	for (size_t i = 0; i < sizeof(gif_screens) / sizeof(gif_screens[0]); i++)
+	{
+		if (rendermode == render_soft)
+		{
+			if (gif_screens[i].owns_pixels)
+			{
+				Z_Free(gif_screens[i].pixels);
+				gif_screens[i].owns_pixels = false;
+			}
+
+			gif_screens[i].size = 0;
+
+			if (i == 1)
+				gif_screens[i].pixels = screens[2];
+			else
+				gif_screens[i].pixels = screens[0];
+		}
+		else
+		{
+			size_t sz = vid.width * vid.height * vid.bpp;
+
+			if (!gif_screens[i].owns_pixels)
+			{
+				gif_screens[i].size = sz;
+				gif_screens[i].pixels = Z_Malloc(gif_screens[i].size, PU_STATIC, NULL);
+				gif_screens[i].owns_pixels = true;
+			}
+			else if (gif_screens[i].size != sz)
+			{
+				gif_screens[i].size = sz;
+				gif_screens[i].pixels = Z_Realloc(gif_screens[i].pixels, gif_screens[i].size, PU_STATIC, NULL);
+			}
+		}
+	}
+}
+
+static void GIF_freescreens(void)
+{
+	for (size_t i = 0; i < sizeof(gif_screens) / sizeof(gif_screens[0]); i++)
+	{
+		if (gif_screens[i].owns_pixels)
+		{
+			Z_Free(gif_screens[i].pixels);
+			gif_screens[i].owns_pixels = false;
+		}
+
+		gif_screens[i].size = 0;
+		gif_screens[i].pixels = NULL;
+	}
+}
+
 //
 // GIF_frame
 // writes a frame into the output gif
 //
 void GIF_frame(void)
 {
-	// there's not much actually needed here, is there.
+	GIF_checkscreens();
 	GIF_framewrite();
 }
 
@@ -781,6 +844,8 @@ INT32 GIF_close(void)
 		Z_Free(giflzw_hashTable);
 	giflzw_hashTable = NULL;
 
+	GIF_freescreens();
+
 	CONS_Printf(M_GetText("Animated gif closed; wrote %d frames\n"), gif_frames);
 	return 1;
 }
