diff --git a/src/p_saveg.c b/src/p_saveg.c
index 650622f59..fe4fed028 100644
--- a/src/p_saveg.c
+++ b/src/p_saveg.c
@@ -5348,7 +5348,7 @@ static inline boolean P_UnArchiveLuabanksAndConsistency(save_t *save_p)
 		case 0x1d: // consistency marker
 			break;
 		default: // anything else is nonsense
-			CONS_Alert(CONS_ERROR, M_GetText("Failed consistency check (???)\n"));
+			CONS_Alert(CONS_ERROR, M_GetText("Failed consistency check\n"));
 			return false;
 	}
 
@@ -5443,9 +5443,9 @@ boolean P_LoadNetGame(save_t *save_p, boolean reloading)
 	// This is stupid and hacky, but maybe it'll work!
 	P_SetRandSeed(P_GetInitSeed());
 
-	// The precipitation would normally be spawned in P_SetupLevel, which is called by
+	// The precipitation would normally be spawned in P_LoadLevel, which is called by
 	// P_NetUnArchiveMisc above. However, that would place it up before P_NetUnArchiveThinkers,
-	// so the thinkers would be deleted later. Therefore, P_SetupLevel will *not* spawn
+	// so the thinkers would be deleted later. Therefore, P_LoadLevel will *not* spawn
 	// precipitation when loading a netgame save. Instead, precip has to be spawned here.
 	// This is done in P_NetUnArchiveSpecials now.
 
