diff --git a/src/w_wad.c b/src/w_wad.c
index 50ba622a9..d02a69e60 100644
--- a/src/w_wad.c
+++ b/src/w_wad.c
@@ -33,6 +33,10 @@
 #include <unistd.h>
 #endif
 
+#if defined(__ANDROID__)
+#include <jni_android.h>
+#endif
+
 #define ZWAD
 
 #ifdef ZWAD
@@ -77,15 +81,26 @@
 #include "hardware/hw_glob.h"
 #endif
 
-#ifdef _DEBUG
 #include "console.h"
-#endif
 
 #ifndef O_BINARY
 #define O_BINARY 0
 #endif
 
+static char filenamebuf[MAX_WADPATH];
+
+// Preemptively store file handles for W_InitFile
+// (File unpacking uses this)
+typedef struct
+{
+	void *handle;
+	char *filename;
+	fhandletype_t type;
+} wadfilehandle_t;
+
+static wadfilehandle_t wadhandles[MAX_WADFILES];
 
+// Lump checklist for file verification
 typedef struct
 {
 	const char *name;
@@ -121,27 +136,485 @@ void W_Shutdown(void)
 	while (numwadfiles--)
 	{
 		wadfile_t *wad = wadfiles[numwadfiles];
+		wadfilehandle_t *wadhandle = &wadhandles[numwadfiles];
 
-		if (wad->handle)
-			fclose(wad->handle);
-		Z_Free(wad->filename);
-		if (wad->path)
-			Z_Free(wad->path);
-		while (wad->numlumps--)
+		if (wadhandle->handle)
 		{
-			if (wad->lumpinfo[wad->numlumps].diskpath)
-				Z_Free(wad->lumpinfo[wad->numlumps].diskpath);
-			Z_Free(wad->lumpinfo[wad->numlumps].longname);
-			Z_Free(wad->lumpinfo[wad->numlumps].fullname);
+			if (wadhandle->handle != wad->handle)
+				File_Close(wadhandle->handle);
+			Z_Free(wadhandle->filename);
 		}
 
-		Z_Free(wad->lumpinfo);
-		Z_Free(wad);
+		W_DeleteResourceFile(wad);
 	}
 
 	Z_Free(wadfiles);
 }
 
+//===========================================================================
+//                                                         MD5 HASH FUNCTIONS
+//===========================================================================
+
+#ifndef NOMD5
+#define MD5_LEN 16
+
+/**
+  * Prints an MD5 string into a human-readable textual format.
+  *
+  * \param md5 The md5 in binary form -- MD5_LEN (16) bytes.
+  * \param buf Where to print the textual form. Needs 2*MD5_LEN+1 (33) bytes.
+  * \author Graue <graue@oceanbase.org>
+  */
+#define MD5_FORMAT \
+	"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x"
+static void PrintMD5String(const UINT8 *md5, char *buf)
+{
+	snprintf(buf, 2*MD5_LEN+1, MD5_FORMAT,
+		md5[0], md5[1], md5[2], md5[3],
+		md5[4], md5[5], md5[6], md5[7],
+		md5[8], md5[9], md5[10], md5[11],
+		md5[12], md5[13], md5[14], md5[15]);
+}
+
+// Convert an md5 string like "7d355827fa8f981482246d6c95f9bd48"
+// into a real md5.
+static void MD5FromString(const char *matchmd5, UINT8 *realmd5)
+{
+	INT32 ix;
+
+	I_Assert(strlen(matchmd5) == 2*MD5_LEN);
+
+	for (ix = 0; ix < 2*MD5_LEN; ix++)
+	{
+		INT32 n, c = matchmd5[ix];
+		if (isdigit(c))
+			n = c - '0';
+		else
+		{
+			I_Assert(isxdigit(c));
+			if (isupper(c)) n = c - 'A' + 10;
+			else n = c - 'a' + 10;
+		}
+		if (ix & 1) realmd5[ix>>1] = (UINT8)(realmd5[ix>>1]+n);
+		else realmd5[ix>>1] = (UINT8)(n<<4);
+	}
+}
+#endif
+
+//===========================================================================
+//                                                             FILE UNPACKING
+//===========================================================================
+
+#ifdef UNPACK_FILES
+static char **startupunpack;
+static UINT16 numstartupunpack = 0;
+static UINT16 startupfiles = 0;
+
+static boolean W_CheckUnpacking(addfilelist_t *list, boolean checkhash);
+static void W_UnpackAlert(alerttype_t level, const char *fmt, ...);
+
+#define UnpackError(err) CONS_Alert(CONS_ERROR, err, __FUNCTION__, filename)
+
+// Unpacks a file into internal storage.
+boolean W_UnpackFile(const char *filename, void *handle)
+{
+	FILE *f;
+	UINT8 buf[UNPACK_BUFFER_SIZE];
+	size_t fullsize, totalread = 0;
+	size_t read, write;
+	boolean success = true;
+
+	INT64 storagespace = 0;
+
+	// progress report
+	unpack_progress_t *progress = &unpack_progress;
+	int status = -1, curstatus;
+	float fstatus;
+
+	if (filename == NULL)
+	{
+		CONS_Alert(CONS_ERROR, "%s: Cannot unpack without a filename.\n", __FUNCTION__);
+		return false;
+	}
+
+	if (handle == NULL)
+	{
+		UnpackError("%s: Cannot unpack %s without a handle.\n");
+		return false;
+	}
+
+	File_Seek(handle, 0, SEEK_END);
+	fullsize = File_Tell(handle);
+	File_Seek(handle, 0, SEEK_SET);
+
+	I_GetDiskFreeSpace(&storagespace);
+	if ((INT64)fullsize > storagespace)
+	{
+		UnpackError("%s: Not enough available storage for caching %s.\n");
+		return false;
+	}
+
+	f = fopen(filename, "w+b");
+	if (!f)
+	{
+		UnpackError("%s: Could not open %s for caching.\n");
+		return false;
+	}
+
+	while (totalread < fullsize)
+	{
+		read = File_Read(buf, 1, UNPACK_BUFFER_SIZE, handle);
+		totalread += read;
+
+		if (progress->report)
+		{
+			fstatus = ((float)totalread / (float)fullsize);
+			curstatus = (int)(fstatus * 100.0f);
+
+			if (curstatus != status)
+			{
+				if (progress->totalfiles)
+				{
+					int diff = (curstatus - status);
+					progress->status += diff;
+					UnpackFile_ProgressReport(min(progress->status / progress->totalfiles, 100));
+				}
+				else
+					UnpackFile_ProgressReport(curstatus);
+
+				status = curstatus;
+			}
+		}
+
+		if (!read)
+			break;
+
+		write = fwrite(buf, 1, read, f);
+		if (write != read)
+		{
+			success = false;
+			break;
+		}
+
+		if (File_EOF(handle))
+			break;
+	}
+
+	fclose(f);
+	return success;
+}
+
+//
+// BASE LIST OF FILES TO UNPACK
+//
+
+static const char *baseunpacklist[] = {
+	"srb2.pk3",
+	"player.dta",
+#ifdef USE_PATCH_DTA
+	"patch.pk3",
+#endif
+	NULL,
+};
+
+static boolean W_CheckInBaseUnpackList(char *filename)
+{
+	INT32 i = 0;
+
+	for (; baseunpacklist[i]; i++)
+	{
+		if (!strcmp(baseunpacklist[i], filename))
+			return true;
+	}
+
+	return false;
+}
+
+void W_UnpackMultipleFiles(addfilelist_t *list, boolean checkhash)
+{
+	W_CheckUnpacking(list, checkhash);
+
+	if (numstartupunpack)
+	{
+		UnpackFile_ProgressClear();
+		UnpackFile_ProgressSetTotalFiles(numstartupunpack);
+		UnpackFile_ProgressSetReportFlag(false);
+		W_UnpackBaseFiles();
+	}
+
+	free(startupunpack);
+}
+
+void W_UnpackBaseFiles(void)
+{
+	UINT16 wadnum = 0;
+	void *handle = NULL, *apkhandle = NULL;
+	const fhandletype_t type = FILEHANDLE_SDL;
+
+	for (; wadnum < startupfiles; wadnum++)
+	{
+		char *filename = startupunpack[wadnum], *fname = filename;
+		wadfilehandle_t *wadhandle = &wadhandles[wadnum];
+
+		if (filename == NULL)
+			continue;
+
+		apkhandle = File_Open(filename, "rb", type);
+
+		if (W_UnpackFile(filename, apkhandle))
+		{
+			// Give a notice that the file was unpacked
+			CONS_Alert(CONS_NOTICE, "Unpacked file %s\n", filename);
+
+			// Open the unpacked file
+			fname = Z_StrDup(va("%s"PATHSEP"%s", I_SystemLocateWad(), filename));
+			handle = File_Open(fname, "rb", type);
+
+			if (handle)
+				File_Close(apkhandle);
+			else
+			{
+				// Couldn't open the unpacked file, load from APK
+				W_UnpackAlert(CONS_WARNING, "Could not open cached %s. Loading will be slower.", filename);
+				handle = apkhandle;
+			}
+		}
+		else
+		{
+			// Couldn't unpack, load from APK
+			W_UnpackAlert(CONS_WARNING, "Could not cache file %s. Loading will be slower.", filename);
+			handle = apkhandle;
+		}
+
+		wadhandle->handle = handle;
+		wadhandle->filename = fname;
+		wadhandle->type = type;
+
+		Z_Free(filename);
+	}
+}
+
+// Checks if a file can be unpacked.
+boolean W_CanUnpackFile(const char *filename, const char *hash, size_t *filesize)
+{
+#if defined(__ANDROID__)
+	void *handle = NULL;
+	char fname[MAX_WADPATH];
+	boolean canunpack = false;
+
+	strncpy(fname, filename, MAX_WADPATH);
+	fname[MAX_WADPATH - 1] = '\0';
+
+	// Check if the specified path contains the file
+	// If it does not, continue checking
+	if ((handle = File_Open(fname, "rb", FILEHANDLE_SDL)) == NULL)
+	{
+		// Remove the path from the filename, leaving only the resource's name itself
+		nameonly(fname);
+
+		// Search through the filesystem
+		// If it was not found, continue checking
+		if (findfile(fname, NULL, true) == FS_NOTFOUND)
+		{
+			handle = File_Open(fname, "rb", FILEHANDLE_SDL);
+
+			if (handle) // If it is found in the application package, it can be unpacked
+			{
+				canunpack = true;
+				if (filesize)
+				{
+					*filesize = File_Size(handle);
+				}
+			}
+		}
+	}
+
+#ifndef NOMD5
+	if (handle && hash && !canunpack)
+	{
+		UINT8 md5sum[16];
+		UINT8 cmpsum[16];
+
+		memset(md5sum, 0x00, 16);
+		memset(cmpsum, 0x00, 16);
+
+		if (!md5_stream_whandle(handle, md5sum))
+		{
+			MD5FromString(hash, cmpsum);
+			if (memcmp(md5sum, cmpsum, 16))
+				canunpack = true;
+		}
+	}
+#else
+	(void)hash;
+#endif
+
+	if (handle)
+		File_Close(handle);
+
+	return canunpack;
+#else
+	(void)filename;
+	(void)hash;
+	return false;
+#endif
+}
+
+static boolean W_CheckUnpacking(addfilelist_t *list, boolean checkhash)
+{
+	size_t totalsize = 0;
+
+	INT64 storagespace = 0;
+	I_GetDiskFreeSpace(&storagespace);
+
+	numstartupunpack = 0;
+	startupunpack = malloc(list->numfiles * sizeof(char*));
+	if (!startupunpack)
+		I_Error("W_CheckUnpacking: out of memory");
+
+	for (UINT16 fnum = 0; fnum < list->numfiles; fnum++)
+	{
+		const char *hash = NULL;
+		size_t size = 0;
+
+		// Get the resource filename
+		// It'll be needed for W_CheckInBaseUnpackList, and for startupunpack[]
+		strncpy(filenamebuf, list->files[fnum], MAX_WADPATH);
+		filenamebuf[MAX_WADPATH - 1] = '\0';
+		nameonly(filenamebuf);
+
+		if (checkhash)
+			hash = list->hashes[fnum];
+
+		if (!W_CheckInBaseUnpackList(filenamebuf) || !W_CanUnpackFile(list->files[fnum], hash, &size))
+			startupunpack[fnum] = NULL;
+		else
+		{
+			startupunpack[fnum] = Z_StrDup(filenamebuf);
+			numstartupunpack++;
+			totalsize += size;
+		}
+
+		startupfiles = fnum;
+	}
+
+	// Not enough storage space
+	if ((INT64)totalsize > storagespace)
+	{
+		CONS_Alert(CONS_ERROR, "Not enough storage space for unpacking files\n");
+		return false;
+	}
+
+	// Startup files can be unpacked
+	return true;
+}
+
+static void W_UnpackAlert(alerttype_t level, const char *fmt, ...)
+{
+	va_list argptr;
+	static char *alert = NULL;
+
+	if (alert == NULL)
+		alert = malloc(8192);
+
+	va_start(argptr, fmt);
+	M_vsnprintf(alert, 8192, fmt, argptr);
+	va_end(argptr);
+
+#if defined(__ANDROID__)
+	JNI_DisplayToast(alert);
+#endif
+
+	CONS_Alert(level, "%s\n", alert);
+}
+
+//
+// PROGRESS REPORTING
+//
+
+unpack_progress_t unpack_progress;
+
+void UnpackFile_ProgressClear(void)
+{
+#ifdef UNPACK_FILES_DEBUG
+	CONS_Printf("UnpackFile_ProgressClear: cleared all progress\n");
+#endif
+	unpack_progress.status = 0;
+	unpack_progress.totalfiles = 0;
+	unpack_progress.report = false;
+}
+
+void UnpackFile_ProgressSetReportFlag(boolean flag)
+{
+#ifdef UNPACK_FILES_DEBUG
+	CONS_Printf("UnpackFile_ProgressSetReportFlag: %d\n", flag);
+#endif
+	unpack_progress.report = flag;
+}
+
+void UnpackFile_ProgressSetTotalFiles(int files)
+{
+#ifdef UNPACK_FILES_DEBUG
+	CONS_Printf("UnpackFile_ProgressSetTotalFiles: file count set to %d\n", files);
+#endif
+	unpack_progress.totalfiles = files;
+}
+
+void UnpackFile_ProgressReport(int progress)
+{
+#ifdef UNPACK_FILES_DEBUG
+	CONS_Printf("UnpackFile_File: %d%% done%s\n", progress, (progress == 100 ? "!" : "..."));
+#endif
+	I_ReportProgress(progress);
+}
+
+#ifdef UNPACK_FILES_DEBUG
+static void UnpackFile_Debug(const char *source, const char *dest)
+{
+	const char *waddir = I_SystemLocateWad();
+	void *handle = File_Open(va("%s"PATHSEP"%s", waddir, source), "rb", FILEHANDLE_SDL);
+
+	if (!handle)
+	{
+		CONS_Alert(CONS_ERROR, "Unpack test failed: couldn't open file %s\n", source);
+		return;
+	}
+
+	if (W_UnpackFile(va("%s"PATHSEP"%s", waddir, dest), handle))
+	{
+		File_Close(handle);
+		handle = File_Open(va("%s"PATHSEP"%s", waddir, dest), "rb", FILEHANDLE_SDL);
+
+		if (handle)
+		{
+			CONS_Alert(CONS_NOTICE, "Unpack test succeeded\n");
+			File_Close(handle);
+		}
+		else
+			CONS_Alert(CONS_ERROR, "Unpack test failed: unpacked file written, but could not open\n");
+	}
+	else
+	{
+		CONS_Alert(CONS_ERROR, "Unpack test failed: couldn't unpack %s\n", source);
+		File_Close(handle);
+	}
+}
+
+void Command_Unpacktest_f(void)
+{
+	UnpackFile_ProgressClear();
+	UnpackFile_ProgressSetTotalFiles(4);
+	UnpackFile_ProgressSetReportFlag(true);
+
+	UnpackFile_Debug("srb2.pk3", "srb2-unpacked.pk3");
+	UnpackFile_Debug("zones.pk3", "zones-unpacked.pk3");
+	UnpackFile_Debug("player.dta", "player-unpacked.dta");
+	UnpackFile_Debug("music.dta", "music-unpacked.dta");
+}
+#endif // UNPACK_FILES_DEBUG
+
+#endif
+
 //===========================================================================
 //                                                        LUMP BASED ROUTINES
 //===========================================================================
@@ -154,16 +627,14 @@ void W_Shutdown(void)
 // Other files are single lumps with the base filename
 //  for the lump name.
 
-static char filenamebuf[MAX_WADPATH];
-
 // W_OpenWadFile
 // Helper function for opening the WAD file.
-// Returns the FILE * handle for the file, or NULL if not found or could not be opened
+// Returns the file handle for the file, or NULL if not found or could not be opened
 // If "useerrors" is true then print errors in the console, else just don't bother
 // "filename" may be modified to have the correct path the actual file is located in, if necessary
-FILE *W_OpenWadFile(const char **filename, boolean useerrors)
+void *W_OpenWadFile(const char **filename, fhandletype_t type, boolean useerrors)
 {
-	FILE *handle;
+	void *handle;
 
 	// Officially, strncpy should not have overlapping buffers, since W_VerifyNMUSlumps is called after this, and it
 	// changes filename to point at filenamebuf, it would technically be doing that. I doubt any issue will occur since
@@ -176,7 +647,7 @@ FILE *W_OpenWadFile(const char **filename, boolean useerrors)
 	}
 
 	// open wad file
-	if ((handle = fopen(*filename, "rb")) == NULL)
+	if ((handle = File_Open(*filename, "rb", type)) == NULL)
 	{
 		// If we failed to load the file with the path as specified by
 		// the user, strip the directories and search for the file.
@@ -186,7 +657,7 @@ FILE *W_OpenWadFile(const char **filename, boolean useerrors)
 		// in filenamebuf == *filename.
 		if (findfile(filenamebuf, NULL, true))
 		{
-			if ((handle = fopen(*filename, "rb")) == NULL)
+			if ((handle = File_Open(*filename, "rb", type)) == NULL)
 			{
 				if (useerrors)
 					CONS_Alert(CONS_ERROR, M_GetText("Can't open %s\n"), *filename);
@@ -195,6 +666,13 @@ FILE *W_OpenWadFile(const char **filename, boolean useerrors)
 		}
 		else
 		{
+#if defined(__ANDROID__)
+			// Search inside the APK
+			handle = File_Open(*filename, "rb", type);
+			if (handle)
+				return handle;
+#endif
+
 			if (useerrors)
 				CONS_Alert(CONS_ERROR, M_GetText("File %s not found.\n"), *filename);
 			return NULL;
@@ -309,28 +787,32 @@ static void W_LoadDehackedLumps(UINT16 wadnum, boolean mainfile)
   * \return 0 if MD5 checksum was made, and is at resblock, 1 if error was found
   */
 
-#ifndef NOMD5
-static INT32 W_MakeFileMD5(const char *filename, void *resblock)
+static INT32 W_MakeFileMD5(const char *filename, fhandletype_t handletype, void *resblock)
 {
-	FILE *fhandle;
+#ifdef NOMD5
+	(void)filename;
+	(void)handletype;
+	memset(resblock, 0x00, 16);
+#else
+	void *fhandle;
 
-	if ((fhandle = fopen(filename, "rb")) != NULL)
+	if ((fhandle = File_Open(filename, "rb", handletype)) != NULL)
 	{
 		tic_t t = I_GetTime();
 		CONS_Debug(DBG_SETUP, "Making MD5 for %s\n",filename);
-		if (md5_stream(fhandle, resblock) == 1)
+		if (md5_stream_whandle(fhandle, resblock) == 1)
 		{
-			fclose(fhandle);
+			File_Close(fhandle);
 			return 1;
 		}
 		CONS_Debug(DBG_SETUP, "MD5 calc for %s took %f seconds\n",
 			filename, (float)(I_GetTime() - t)/NEWTICRATE);
-		fclose(fhandle);
+		File_Close(fhandle);
 		return 0;
 	}
+#endif
 	return 1;
 }
-#endif
 
 // Invalidates the cache of lump numbers. Call this whenever a wad is added.
 static void W_InvalidateLumpnumCache(void)
@@ -338,30 +820,57 @@ static void W_InvalidateLumpnumCache(void)
 	memset(lumpnumcache, 0, sizeof (lumpnumcache));
 }
 
+static boolean MagicIsWAD(char id[4])
+{
+	// Very likely a wad
+	if (!memcmp(id, "IWAD", 4) || !memcmp(id, "PWAD", 4) || !memcmp(id, "ZWAD", 4) || !memcmp(id, "SDLL", 4))
+		return true;
+
+	return false;
+}
+
 /** Detect a file type.
- * \todo Actually detect the wad/pkzip headers and whatnot, instead of just checking the extensions.
  */
-static restype_t ResourceFileDetect (const char* filename)
+static restype_t ResourceFileDetect (filehandle_t *handle, const char* filename)
 {
-	if (!stricmp(&filename[strlen(filename) - 4], ".pk3"))
+	char id[4];
+	size_t read;
+
+	// Read the first four bytes, then seek back
+	read = File_Read(&id, 1, sizeof id, handle);
+
+	File_Seek(handle, 0, SEEK_SET);
+
+	if (read >= sizeof id)
+	{
+		if (MagicIsWAD(id))
+			return RET_WAD;
+		// Seems to be a zip (so, a pk3)
+		else if (!memcmp(id, "PK\x03\x04", 4))
+			return RET_PK3;
+	}
+
+	// Couldn't figure it out, let's just look at the filename
+	if (!stricmp(&filename[strlen(filename) - 4], ".pk3") || !stricmp(&filename[strlen(filename) - 4], ".zip"))
 		return RET_PK3;
 	if (!stricmp(&filename[strlen(filename) - 4], ".soc"))
 		return RET_SOC;
 	if (!stricmp(&filename[strlen(filename) - 4], ".lua"))
 		return RET_LUA;
 
+	// I give up! Assume it's a WAD
 	return RET_WAD;
 }
 
 /** Create a 1-lump lumpinfo_t for standalone files.
  */
-static lumpinfo_t* ResGetLumpsStandalone (FILE* handle, UINT16* numlumps, const char* lumpname)
+static lumpinfo_t* ResGetLumpsStandalone (void* handle, UINT16* numlumps, const char* lumpname)
 {
 	lumpinfo_t* lumpinfo = Z_Calloc(sizeof (*lumpinfo), PU_STATIC, NULL);
 	lumpinfo->position = 0;
-	fseek(handle, 0, SEEK_END);
-	lumpinfo->size = ftell(handle);
-	fseek(handle, 0, SEEK_SET);
+	File_Seek(handle, 0, SEEK_END);
+	lumpinfo->size = File_Tell(handle);
+	File_Seek(handle, 0, SEEK_SET);
 	strcpy(lumpinfo->name, lumpname);
 	lumpinfo->hash = quickncasehash(lumpname, 8);
 
@@ -381,7 +890,7 @@ static lumpinfo_t* ResGetLumpsStandalone (FILE* handle, UINT16* numlumps, const
 
 /** Create a lumpinfo_t array for a WAD file.
  */
-static lumpinfo_t* ResGetLumpsWad (FILE* handle, UINT16* nlmp, const char* filename)
+static lumpinfo_t* ResGetLumpsWad (void* handle, UINT16* nlmp, const char* filename)
 {
 	UINT16 numlumps = *nlmp;
 	lumpinfo_t* lumpinfo;
@@ -394,17 +903,15 @@ static lumpinfo_t* ResGetLumpsWad (FILE* handle, UINT16* nlmp, const char* filen
 	void *fileinfov;
 
 	// read the header
-	if (fread(&header, 1, sizeof header, handle) < sizeof header)
+	if (File_Read(&header, 1, sizeof header, handle) < sizeof header)
 	{
-		CONS_Alert(CONS_ERROR, M_GetText("Can't read wad header because %s\n"), M_FileError(handle));
+		CONS_Alert(CONS_ERROR, M_GetText("Can't read wad header because %s\n"), File_Error(handle));
 		return NULL;
 	}
 
 	if (memcmp(header.identification, "ZWAD", 4) == 0)
 		compressed = 1;
-	else if (memcmp(header.identification, "IWAD", 4) != 0
-		&& memcmp(header.identification, "PWAD", 4) != 0
-		&& memcmp(header.identification, "SDLL", 4) != 0)
+	else if (!MagicIsWAD(header.identification))
 	{
 		CONS_Alert(CONS_ERROR, M_GetText("Invalid WAD header\n"));
 		return NULL;
@@ -416,10 +923,10 @@ static lumpinfo_t* ResGetLumpsWad (FILE* handle, UINT16* nlmp, const char* filen
 	// read wad file directory
 	i = header.numlumps * sizeof (*fileinfo);
 	fileinfov = fileinfo = malloc(i);
-	if (fseek(handle, header.infotableofs, SEEK_SET) == -1
-		|| fread(fileinfo, 1, i, handle) < i)
+	if (File_Seek(handle, header.infotableofs, SEEK_SET) == -1
+		|| File_Read(fileinfo, 1, i, handle) < i)
 	{
-		CONS_Alert(CONS_ERROR, M_GetText("Corrupt wadfile directory (%s)\n"), M_FileError(handle));
+		CONS_Alert(CONS_ERROR, M_GetText("Corrupt wadfile directory (%s)\n"), File_Error(handle));
 		free(fileinfov);
 		return NULL;
 	}
@@ -436,12 +943,12 @@ static lumpinfo_t* ResGetLumpsWad (FILE* handle, UINT16* nlmp, const char* filen
 		if (compressed) // wad is compressed, lump might be
 		{
 			UINT32 realsize = 0;
-			if (fseek(handle, lump_p->position, SEEK_SET)
-				== -1 || fread(&realsize, 1, sizeof realsize,
+			if (File_Seek(handle, lump_p->position, SEEK_SET)
+				== -1 || File_Read(&realsize, 1, sizeof realsize,
 				handle) < sizeof realsize)
 			{
 				I_Error("corrupt compressed file: %s; maybe %s", /// \todo Avoid the bailout?
-					filename, M_FileError(handle));
+					filename, File_Error(handle));
 			}
 			realsize = LONG(realsize);
 			if (realsize != 0)
@@ -481,15 +988,24 @@ static lumpinfo_t* ResGetLumpsWad (FILE* handle, UINT16* nlmp, const char* filen
 
 /** Optimized pattern search in a file.
  */
-static boolean ResFindSignature (FILE* handle, char endPat[], UINT32 startpos)
+static boolean ResFindSignature (void* handle, char endPat[], UINT32 startpos)
 {
 	char *s;
 	int c;
 
-	fseek(handle, startpos, SEEK_SET);
+	File_Seek(handle, startpos, SEEK_SET);
 	s = endPat;
-	while((c = fgetc(handle)) != EOF)
+
+#if defined(__ANDROID__)
+	while (true)
+	{
+		c = (unsigned char)(File_GetChar(handle));
+		if (File_EOF(handle))
+			break;
+#else
+	while((c = File_GetChar(handle)) != EOF)
 	{
+#endif
 		if (*s != c && s > endPat) // No match?
 			s = endPat; // We "reset" the counter by sending the s pointer back to the start of the array.
 		if (*s == c)
@@ -560,7 +1076,7 @@ typedef struct zlentry_s
 
 /** Create a lumpinfo_t array for a PKZip file.
  */
-static lumpinfo_t* ResGetLumpsZip (FILE* handle, UINT16* nlmp)
+static lumpinfo_t* ResGetLumpsZip (void* handle, UINT16* nlmp)
 {
     zend_t zend;
     zentry_t zentry;
@@ -576,33 +1092,33 @@ static lumpinfo_t* ResGetLumpsZip (FILE* handle, UINT16* nlmp)
 
 	// Look for central directory end signature near end of file.
 	// Contains entry number (number of lumps), and central directory start offset.
-	fseek(handle, 0, SEEK_END);
-	if (!ResFindSignature(handle, pat_end, max(0, ftell(handle) - (22 + 65536))))
+	File_Seek(handle, 0, SEEK_END);
+	if (!ResFindSignature(handle, pat_end, max(0, File_Tell(handle) - (22 + 65536))))
 	{
 		CONS_Alert(CONS_ERROR, "Missing central directory\n");
 		return NULL;
 	}
 
-	fseek(handle, -4, SEEK_CUR);
-	if (fread(&zend, 1, sizeof zend, handle) < sizeof zend)
+	File_Seek(handle, -4, SEEK_CUR);
+	if (File_Read(&zend, 1, sizeof zend, handle) < sizeof zend)
 	{
-		CONS_Alert(CONS_ERROR, "Corrupt central directory (%s)\n", M_FileError(handle));
+		CONS_Alert(CONS_ERROR, "Corrupt central directory (%s)\n", File_Error(handle));
 		return NULL;
 	}
 	numlumps = zend.entries;
 
 	lump_p = lumpinfo = Z_Malloc(numlumps * sizeof (*lumpinfo), PU_STATIC, NULL);
 
-	fseek(handle, zend.cdiroffset, SEEK_SET);
+	File_Seek(handle, zend.cdiroffset, SEEK_SET);
 	for (i = 0; i < numlumps; i++, lump_p++)
 	{
 		char* fullname;
 		char* trimname;
 		char* dotpos;
 
-		if (fread(&zentry, 1, sizeof(zentry_t), handle) < sizeof(zentry_t))
+		if (File_Read(&zentry, 1, sizeof(zentry_t), handle) < sizeof(zentry_t))
 		{
-			CONS_Alert(CONS_ERROR, "Failed to read central directory (%s)\n", M_FileError(handle));
+			CONS_Alert(CONS_ERROR, "Failed to read central directory (%s)\n", File_Error(handle));
 			Z_Free(lumpinfo);
 			return NULL;
 		}
@@ -619,9 +1135,9 @@ static lumpinfo_t* ResGetLumpsZip (FILE* handle, UINT16* nlmp)
 		lump_p->size = zentry.size;
 
 		fullname = malloc(zentry.namelen + 1);
-		if (fgets(fullname, zentry.namelen + 1, handle) != fullname)
+		if (File_GetString(fullname, zentry.namelen + 1, handle) != fullname)
 		{
-			CONS_Alert(CONS_ERROR, "Unable to read lumpname (%s)\n", M_FileError(handle));
+			CONS_Alert(CONS_ERROR, "Unable to read lumpname (%s)\n", File_Error(handle));
 			Z_Free(lumpinfo);
 			free(fullname);
 			return NULL;
@@ -668,7 +1184,7 @@ static lumpinfo_t* ResGetLumpsZip (FILE* handle, UINT16* nlmp)
 		free(fullname);
 
 		// skip and ignore comments/extra fields
-		if (fseek(handle, zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
+		if (File_Seek(handle, zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
 		{
 			CONS_Alert(CONS_ERROR, "Central directory is corrupt\n");
 			Z_Free(lumpinfo);
@@ -680,7 +1196,7 @@ static lumpinfo_t* ResGetLumpsZip (FILE* handle, UINT16* nlmp)
 	for (i = 0, lump_p = lumpinfo; i < numlumps; i++, lump_p++)
 	{
 		// skip and ignore comments/extra fields
-		if ((fseek(handle, lump_p->position, SEEK_SET) != 0) || (fread(&zlentry, 1, sizeof(zlentry_t), handle) < sizeof(zlentry_t)))
+		if ((File_Seek(handle, lump_p->position, SEEK_SET) != 0) || (File_Read(&zlentry, 1, sizeof(zlentry_t), handle) < sizeof(zlentry_t)))
 		{
 			CONS_Alert(CONS_ERROR, "Local headers for lump %s are corrupt\n", lump_p->fullname);
 			Z_Free(lumpinfo);
@@ -848,11 +1364,12 @@ static void W_LoadTrnslateLumps(UINT16 w)
 //
 // Can now load dehacked files (.soc)
 //
-UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
+UINT16 W_InitFile(const char *filename, fhandletype_t handletype, boolean mainfile, boolean startup)
 {
-	FILE *handle;
+	void *handle;
 	lumpinfo_t *lumpinfo = NULL;
 	wadfile_t *wadfile;
+	wadfilehandle_t *wadhandle;
 	restype_t type;
 	UINT16 numlumps = 0;
 #ifndef NOMD5
@@ -885,14 +1402,22 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 	}
 
 	// open wad file
-	if ((handle = W_OpenWadFile(&filename, true)) == NULL)
+	wadhandle = &wadhandles[numwadfiles];
+	if (wadhandle->handle)
+	{
+		handle = wadhandle->handle;
+		handletype = wadhandle->type;
+		filename = wadhandle->filename;
+	}
+	else if ((handle = W_OpenWadFile(&filename, handletype, true)) == NULL)
 		return W_InitFileError(filename, startup);
 
-	important = W_VerifyNMUSlumps(filename, startup);
+	important = W_VerifyNMUSlumps(filename, handletype, startup);
 
 	if (important == -1)
 	{
-		fclose(handle);
+		if (handle != wadhandle->handle)
+			File_Close(handle);
 		return INT16_MAX;
 	}
 
@@ -904,7 +1429,7 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 	// Let's not add a wad file if the MD5 matches
 	// an MD5 of an already added WAD file!
 	//
-	W_MakeFileMD5(filename, md5sum);
+	W_MakeFileMD5(filename, handletype, md5sum);
 
 	for (i = 0; i < numwadfiles; i++)
 	{
@@ -915,13 +1440,13 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 		{
 			CONS_Alert(CONS_ERROR, M_GetText("%s is already loaded\n"), filename);
 			if (handle)
-				fclose(handle);
+				File_Close(handle);
 			return W_InitFileError(filename, false);
 		}
 	}
 #endif
 
-	switch(type = ResourceFileDetect(filename))
+	switch(type = ResourceFileDetect(handle, filename))
 	{
 	case RET_SOC:
 		lumpinfo = ResGetLumpsStandalone(handle, &numlumps, "OBJCTCFG");
@@ -941,7 +1466,7 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 
 	if (lumpinfo == NULL)
 	{
-		fclose(handle);
+		File_Close(handle);
 		return W_InitFileError(filename, startup);
 	}
 
@@ -963,9 +1488,8 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 	wadfile->foldercount = 0;
 	wadfile->lumpinfo = lumpinfo;
 	wadfile->important = important;
-	fseek(handle, 0, SEEK_END);
-	wadfile->filesize = (unsigned)ftell(handle);
-	wadfile->type = type;
+	File_Seek(handle, 0, SEEK_END);
+	wadfile->filesize = (unsigned)File_Tell(handle);
 
 	// already generated, just copy it over
 	M_Memcpy(&wadfile->md5sum, &md5sum, 16);
@@ -1018,6 +1542,136 @@ UINT16 W_InitFile(const char *filename, boolean mainfile, boolean startup)
 	return wadfile->numlumps;
 }
 
+wadfile_t *W_LoadResourceFile(const char *filename, fhandletype_t handletype)
+{
+	void *handle;
+	lumpinfo_t *lumpinfo = NULL;
+	wadfile_t *wadfile;
+	restype_t type;
+	UINT16 numlumps = 0;
+
+	// open wad file
+	if ((handle = File_Open(filename, "rb", handletype)) == NULL)
+	{
+		CONS_Printf(M_GetText("Errors occurred while loading %s.\n"), filename);
+		return NULL;
+	}
+
+	switch (type = ResourceFileDetect(handle, filename))
+	{
+	case RET_PK3:
+		lumpinfo = ResGetLumpsZip(handle, &numlumps);
+		break;
+	case RET_WAD:
+		lumpinfo = ResGetLumpsWad(handle, &numlumps, filename);
+		break;
+	default:
+		CONS_Alert(CONS_ERROR, "Unsupported file format\n");
+	}
+
+	if (lumpinfo == NULL)
+	{
+		File_Close(handle);
+		CONS_Printf(M_GetText("Errors occurred while loading %s.\n"), filename);
+		return NULL;
+	}
+
+	wadfile = Z_Malloc(sizeof (*wadfile), PU_STATIC, NULL);
+	wadfile->filename = Z_StrDup(filename);
+	wadfile->path = NULL;
+	wadfile->type = type;
+	wadfile->handle = handle;
+	wadfile->numlumps = numlumps;
+	wadfile->foldercount = 0;
+	wadfile->lumpinfo = lumpinfo;
+	wadfile->important = false;
+	File_Seek(handle, 0, SEEK_END);
+	wadfile->filesize = (unsigned)File_Tell(handle);
+
+	// Irrelevant.
+	memset(wadfile->md5sum, 0x00, 16);
+
+	Z_Calloc(numlumps * sizeof (*wadfile->lumpcache), PU_STATIC, &wadfile->lumpcache);
+	Z_Calloc(numlumps * sizeof (*wadfile->patchcache), PU_STATIC, &wadfile->patchcache);
+
+	return wadfile;
+}
+
+void W_DeleteResourceFile(wadfile_t *wad)
+{
+	if (!wad)
+		return;
+
+	if (wad->handle)
+		File_Close(wad->handle);
+	Z_Free(wad->filename);
+	if (wad->path)
+		Z_Free(wad->path);
+
+	while (wad->numlumps--)
+	{
+		Z_Free(wad->lumpcache[wad->numlumps]);
+		if (wad->patchcache[wad->numlumps])
+			Patch_Free(wad->patchcache[wad->numlumps]);
+		if (wad->lumpinfo[wad->numlumps].diskpath)
+			Z_Free(wad->lumpinfo[wad->numlumps].diskpath);
+		Z_Free(wad->lumpinfo[wad->numlumps].longname);
+		Z_Free(wad->lumpinfo[wad->numlumps].fullname);
+	}
+
+	Z_Free(wad->lumpcache);
+	Z_Free(wad->patchcache);
+	Z_Free(wad->lumpinfo);
+	Z_Free(wad);
+}
+
+UINT16 Resource_CheckNumForName(wadfile_t *wad, const char *name)
+{
+	lumpinfo_t *lump_p = wad->lumpinfo;
+	for (UINT16 i = 0; i < wad->numlumps; i++, lump_p++)
+		if (!strcmp(lump_p->fullname, name))
+			return i;
+
+	// not found.
+	return INT16_MAX;
+}
+
+void *Resource_CacheLumpNum(wadfile_t *wad, UINT16 lump, INT32 tag)
+{
+	lumpcache_t *lumpcache;
+
+	if (lump >= wad->numlumps)
+		return NULL;
+
+	lumpcache = wad->lumpcache;
+	if (!lumpcache[lump])
+	{
+		void *ptr = Z_Malloc(Resource_LumpLength(wad, lump), tag, &lumpcache[lump]);
+		Resource_ReadLumpHeader(wad, lump, ptr, 0, 0);  // read the lump in full
+	}
+	else
+		Z_ChangeTag(lumpcache[lump], tag);
+
+	return lumpcache[lump];
+}
+
+void *Resource_CacheLumpName(wadfile_t *wad, const char *name, INT32 tag)
+{
+	UINT16 lumpnum = Resource_CheckNumForName(wad, name);
+	if (lumpnum == INT16_MAX)
+	{
+		CONS_Alert(CONS_ERROR, "Resource file %s does not contain any lump named %s\n", wad->filename, name);
+		return NULL;
+	}
+
+	return Resource_CacheLumpNum(wad, lumpnum, tag);
+}
+
+boolean Resource_LumpExists(wadfile_t *wad, const char *name)
+{
+	return Resource_CheckNumForName(wad, name) != INT16_MAX;
+}
+
 //
 // Loads a folder as a WAD.
 //
@@ -1193,7 +1847,7 @@ UINT16 W_InitFolder(const char *path, boolean mainfile, boolean startup)
   *
   * \param filenames A null-terminated list of files to use.
   */
-void W_InitMultipleFiles(addfilelist_t *list)
+void W_InitMultipleFiles(addfilelist_t *list, fhandletype_t handletype)
 {
 	size_t i = 0;
 
@@ -1208,7 +1862,17 @@ void W_InitMultipleFiles(addfilelist_t *list)
 		if (pathsep == '\\' || pathsep == '/')
 			W_InitFolder(fn, mainfile, true);
 		else
-			W_InitFile(fn, mainfile, true);
+		{
+			UINT16 status = W_InitFile(fn, handletype, mainfile, true);
+
+#ifndef DEVELOP
+			// Check MD5s of autoloaded files
+			if (mainfile && status != INT16_MAX && list->hashes[i])
+				W_VerifyFileMD5(numwadfiles - 1, list->hashes[i]);
+#else
+			(void)status;
+#endif
+		}
 	}
 }
 
@@ -1900,17 +2564,17 @@ UINT8 W_LumpExists(const char *name)
 	return false;
 }
 
-size_t W_LumpLengthPwad(UINT16 wad, UINT16 lump)
+size_t Resource_LumpLength(wadfile_t *wad, UINT16 lump)
 {
 	lumpinfo_t *l;
 
-	if (!TestValidLump(wad, lump))
+	if (lump >= wad->numlumps)
 		return 0;
 
-	l = wadfiles[wad]->lumpinfo + lump;
+	l = wad->lumpinfo + lump;
 
 	// Open the external file for this lump, if the WAD is a folder.
-	if (wadfiles[wad]->type == RET_FOLDER)
+	if (wad->type == RET_FOLDER)
 	{
 		// pathisdirectory calls stat, so if anything wrong has happened,
 		// this is the time to be aware of it.
@@ -1944,6 +2608,14 @@ size_t W_LumpLengthPwad(UINT16 wad, UINT16 lump)
 	return l->size;
 }
 
+size_t W_LumpLengthPwad(UINT16 wad, UINT16 lump)
+{
+	if (!TestValidLump(wad, lump))
+		return 0;
+
+	return Resource_LumpLength(wadfiles[wad], lump);
+}
+
 /** Returns the buffer size needed to load the given lump.
   *
   * \param lump Lump number to look at.
@@ -2030,19 +2702,19 @@ void zerr(int ret)
   * \return Number of bytes read (should equal size).
   * \sa W_ReadLump, W_RawReadLumpHeader
   */
-size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, size_t offset)
+size_t Resource_ReadLumpHeader(wadfile_t *wad, UINT16 lump, void *dest, size_t size, size_t offset)
 {
 	size_t lumpsize, bytesread;
 	lumpinfo_t *l;
-	FILE *handle = NULL;
+	void *handle = NULL;
 
-	if (!TestValidLump(wad, lump))
+	if (lump >= wad->numlumps)
 		return 0;
 
-	l = wadfiles[wad]->lumpinfo + lump;
+	l = wad->lumpinfo + lump;
 
 	// Open the external file for this lump, if the WAD is a folder.
-	if (wadfiles[wad]->type == RET_FOLDER)
+	if (wad->type == RET_FOLDER)
 	{
 		// pathisdirectory calls stat, so if anything wrong has happened,
 		// this is the time to be aware of it.
@@ -2052,33 +2724,33 @@ size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, si
 		{
 #ifndef AVOID_ERRNO
 			if (direrror == ENOENT)
-				I_Error("W_ReadLumpHeaderPwad: file %s doesn't exist", l->diskpath);
+				I_Error("Resource_ReadLumpHeader: file %s doesn't exist", l->diskpath);
 			else
-				I_Error("W_ReadLumpHeaderPwad: could not stat %s: %s", l->diskpath, strerror(direrror));
+				I_Error("Resource_ReadLumpHeader: could not stat %s: %s", l->diskpath, strerror(direrror));
 #else
-			I_Error("W_ReadLumpHeaderPwad: could not access %s", l->diskpath);
+			I_Error("Resource_ReadLumpHeader: could not access %s", l->diskpath);
 #endif
 		}
 		else if (stat == 1) // Path is a folder.
 			return 0;
 		else
 		{
-			handle = fopen(l->diskpath, "rb");
+			handle = File_Open(l->diskpath, "rb", FILEHANDLE_STANDARD);
 			if (handle == NULL)
-				I_Error("W_ReadLumpHeaderPwad: could not open file %s", l->diskpath);
+				I_Error("Resource_ReadLumpHeader: could not open file %s", l->diskpath);
 
 			// Find length of file
-			fseek(handle, 0, SEEK_END);
-			l->size = l->disksize = ftell(handle);
+			File_Seek(handle, 0, SEEK_END);
+			l->size = l->disksize = File_Tell(handle);
 		}
 	}
 
-	lumpsize = wadfiles[wad]->lumpinfo[lump].size;
+	lumpsize = wad->lumpinfo[lump].size;
 	// empty resource (usually markers like S_START, F_END ..)
 	if (!lumpsize || lumpsize<offset)
 	{
-		if (wadfiles[wad]->type == RET_FOLDER)
-			fclose(handle);
+		if (wad->type == RET_FOLDER)
+			File_Close(handle);
 		return 0;
 	}
 
@@ -2088,17 +2760,17 @@ size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, si
 
 	// Let's get the raw lump data.
 	// We setup the desired file handle to read the lump data.
-	if (wadfiles[wad]->type != RET_FOLDER)
-		handle = wadfiles[wad]->handle;
-	fseek(handle, (long)(l->position + offset), SEEK_SET);
+	if (wad->type != RET_FOLDER)
+		handle = wad->handle;
+	File_Seek(handle, (long)(l->position + offset), SEEK_SET);
 
 	// But let's not copy it yet. We support different compression formats on lumps, so we need to take that into account.
-	switch(wadfiles[wad]->lumpinfo[lump].compression)
+	switch(wad->lumpinfo[lump].compression)
 	{
 	case CM_NOCOMPRESSION:		// If it's uncompressed, we directly write the data into our destination, and return the bytes read.
-		bytesread = fread(dest, 1, size, handle);
-		if (wadfiles[wad]->type == RET_FOLDER)
-			fclose(handle);
+		bytesread = File_Read(dest, 1, size, handle);
+		if (wad->type == RET_FOLDER)
+			File_Close(handle);
 		return bytesread;
 	case CM_LZF:		// Is it LZF compressed? Used by ZWADs.
 		{
@@ -2110,23 +2782,23 @@ size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, si
 			rawData = Z_Malloc(l->disksize, PU_STATIC, NULL);
 			decData = Z_Malloc(l->size, PU_STATIC, NULL);
 
-			if (fread(rawData, 1, l->disksize, handle) < l->disksize)
-				I_Error("wad %d, lump %d: cannot read compressed data", wad, lump);
+			if (File_Read(rawData, 1, l->disksize, handle) < l->disksize)
+				I_Error("wad %s, lump %d: cannot read compressed data", wad->filename, lump);
 			retval = lzf_decompress(rawData, l->disksize, decData, l->size);
 #ifndef AVOID_ERRNO
 			if (retval == 0) // If this was returned, check if errno was set
 			{
 				// errno is a global var set by the lzf functions when something goes wrong.
 				if (errno == E2BIG)
-					I_Error("wad %d, lump %d: compressed data too big (bigger than %s)", wad, lump, sizeu1(l->size));
+					I_Error("wad %s, lump %d: compressed data too big (bigger than %s)", wad->filename, lump, sizeu1(l->size));
 				else if (errno == EINVAL)
-					I_Error("wad %d, lump %d: invalid compressed data", wad, lump);
+					I_Error("wad %s, lump %d: invalid compressed data", wad->filename, lump);
 			}
 			// Otherwise, fall back on below error (if zero was actually the correct size then ???)
 #endif
 			if (retval != l->size)
 			{
-				I_Error("wad %d, lump %d: decompressed to wrong number of bytes (expected %s, got %s)", wad, lump, sizeu1(l->size), sizeu2(retval));
+				I_Error("wad %s, lump %d: decompressed to wrong number of bytes (expected %s, got %s)", wad->filename, lump, sizeu1(l->size), sizeu2(retval));
 			}
 
 			if (!decData) // Did we get no data at all?
@@ -2155,8 +2827,8 @@ size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, si
 			rawData = Z_Malloc(rawSize, PU_STATIC, NULL);
 			decData = Z_Malloc(decSize, PU_STATIC, NULL);
 
-			if (fread(rawData, 1, rawSize, handle) < rawSize)
-				I_Error("wad %d, lump %d: cannot read compressed data", wad, lump);
+			if (File_Read(rawData, 1, rawSize, handle) < rawSize)
+				I_Error("wad %s, lump %d: cannot read compressed data", wad->filename, lump);
 
 			strm.zalloc = Z_NULL;
 			strm.zfree = Z_NULL;
@@ -2197,11 +2869,19 @@ size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, si
 		}
 #endif
 	default:
-		I_Error("wad %d, lump %d: unsupported compression type!", wad, lump);
+		I_Error("wad %s, lump %d: unsupported compression type!", wad->filename, lump);
 	}
 	return 0;
 }
 
+size_t W_ReadLumpHeaderPwad(UINT16 wad, UINT16 lump, void *dest, size_t size, size_t offset)
+{
+	if (!TestValidLump(wad, lump))
+		return 0;
+
+	return Resource_ReadLumpHeader(wadfiles[wad], lump, dest, size, offset);
+}
+
 size_t W_ReadLumpHeader(lumpnum_t lumpnum, void *dest, size_t size, size_t offset)
 {
 	return W_ReadLumpHeaderPwad(WADFILENUM(lumpnum), LUMPNUM(lumpnum), dest, size, offset);
@@ -2229,21 +2909,10 @@ void W_ReadLumpPwad(UINT16 wad, UINT16 lump, void *dest)
 // ==========================================================================
 void *W_CacheLumpNumPwad(UINT16 wad, UINT16 lump, INT32 tag)
 {
-	lumpcache_t *lumpcache;
-
 	if (!TestValidLump(wad,lump))
 		return NULL;
 
-	lumpcache = wadfiles[wad]->lumpcache;
-	if (!lumpcache[lump])
-	{
-		void *ptr = Z_Malloc(W_LumpLengthPwad(wad, lump), tag, &lumpcache[lump]);
-		W_ReadLumpHeaderPwad(wad, lump, ptr, 0, 0);  // read the lump in full
-	}
-	else
-		Z_ChangeTag(lumpcache[lump], tag);
-
-	return lumpcache[lump];
+	return Resource_CacheLumpNum(wadfiles[wad], lump, tag);
 }
 
 void *W_CacheLumpNum(lumpnum_t lumpnum, INT32 tag)
@@ -2516,27 +3185,19 @@ void *W_CachePatchLongName(const char *name, INT32 tag)
 		return W_CachePatchNum(W_GetNumForLongPatchName("MISSING"), tag);
 	return W_CachePatchNum(num, tag);
 }
-#ifndef NOMD5
-#define MD5_LEN 16
-
-/**
-  * Prints an MD5 string into a human-readable textual format.
-  *
-  * \param md5 The md5 in binary form -- MD5_LEN (16) bytes.
-  * \param buf Where to print the textual form. Needs 2*MD5_LEN+1 (33) bytes.
-  * \author Graue <graue@oceanbase.org>
-  */
-#define MD5_FORMAT \
-	"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x"
-static void PrintMD5String(const UINT8 *md5, char *buf)
+#if defined(__ANDROID__)
+static boolean W_IsAndroidPK3(const char *filename)
 {
-	snprintf(buf, 2*MD5_LEN+1, MD5_FORMAT,
-		md5[0], md5[1], md5[2], md5[3],
-		md5[4], md5[5], md5[6], md5[7],
-		md5[8], md5[9], md5[10], md5[11],
-		md5[12], md5[13], md5[14], md5[15]);
+	char androidpk3[MAX_WADPATH];
+
+	strncpy(androidpk3, filename, MAX_WADPATH);
+	androidpk3[MAX_WADPATH - 1] = '\0';
+	nameonly(androidpk3);
+
+	return (!strcmp(androidpk3, ANDROID_PK3_FILENAME));
 }
 #endif
+
 /** Verifies a file's MD5 is as it should be.
   * For releases, used as cheat prevention -- if the MD5 doesn't match, a
   * fatal error is thrown. In debug mode, an MD5 mismatch only triggers a
@@ -2554,26 +3215,10 @@ void W_VerifyFileMD5(UINT16 wadfilenum, const char *matchmd5)
 	(void)matchmd5;
 #else
 	UINT8 realmd5[MD5_LEN];
-	INT32 ix;
 
-	I_Assert(strlen(matchmd5) == 2*MD5_LEN);
 	I_Assert(wadfilenum < numwadfiles);
-	// Convert an md5 string like "7d355827fa8f981482246d6c95f9bd48"
-	// into a real md5.
-	for (ix = 0; ix < 2*MD5_LEN; ix++)
-	{
-		INT32 n, c = matchmd5[ix];
-		if (isdigit(c))
-			n = c - '0';
-		else
-		{
-			I_Assert(isxdigit(c));
-			if (isupper(c)) n = c - 'A' + 10;
-			else n = c - 'a' + 10;
-		}
-		if (ix & 1) realmd5[ix>>1] = (UINT8)(realmd5[ix>>1]+n);
-		else realmd5[ix>>1] = (UINT8)(n<<4);
-	}
+
+	MD5FromString(matchmd5, realmd5);
 
 	if (memcmp(realmd5, wadfiles[wadfilenum]->md5sum, 16))
 	{
@@ -2608,7 +3253,7 @@ W_VerifyName (const char *name, lumpchecklist_t *checklist, boolean status)
 }
 
 static int
-W_VerifyWAD (FILE *fp, lumpchecklist_t *checklist, boolean status)
+W_VerifyWAD (void *fp, lumpchecklist_t *checklist, boolean status)
 {
 	size_t i;
 
@@ -2617,7 +3262,7 @@ W_VerifyWAD (FILE *fp, lumpchecklist_t *checklist, boolean status)
 	filelump_t lumpinfo;
 
 	// read the header
-	if (fread(&header, 1, sizeof header, fp) == sizeof header
+	if (File_Read(&header, 1, sizeof header, fp) == sizeof header
 			&& header.numlumps < INT16_MAX
 			&& strncmp(header.identification, "ZWAD", 4)
 			&& strncmp(header.identification, "IWAD", 4)
@@ -2631,13 +3276,13 @@ W_VerifyWAD (FILE *fp, lumpchecklist_t *checklist, boolean status)
 	header.infotableofs = LONG(header.infotableofs);
 
 	// let seek to the lumpinfo list
-	if (fseek(fp, header.infotableofs, SEEK_SET) == -1)
+	if (File_Seek(fp, header.infotableofs, SEEK_SET) == -1)
 		return true;
 
 	for (i = 0; i < header.numlumps; i++)
 	{
 		// fill in lumpinfo for this wad file directory
-		if (fread(&lumpinfo, sizeof (lumpinfo), 1 , fp) != 1)
+		if (File_Read(&lumpinfo, sizeof (lumpinfo), 1 , fp) != 1)
 			return true;
 
 		lumpinfo.filepos = LONG(lumpinfo.filepos);
@@ -2668,7 +3313,7 @@ static lumpchecklist_t folderblacklist[] =
 };
 
 static int
-W_VerifyPK3 (FILE *fp, lumpchecklist_t *checklist, boolean status)
+W_VerifyPK3 (void *fp, lumpchecklist_t *checklist, boolean status)
 {
 	int verified = true;
 
@@ -2694,28 +3339,28 @@ W_VerifyPK3 (FILE *fp, lumpchecklist_t *checklist, boolean status)
 
 	// Central directory bullshit
 
-	fseek(fp, 0, SEEK_END);
-	file_size = ftell(fp);
+	File_Seek(fp, 0, SEEK_END);
+	file_size = File_Tell(fp);
 
-	if (!ResFindSignature(fp, pat_end, max(0, ftell(fp) - (22 + 65536))))
+	if (!ResFindSignature(fp, pat_end, max(0, File_Tell(fp) - (22 + 65536))))
 		return true;
 
-	fseek(fp, -4, SEEK_CUR);
-	if (fread(&zend, 1, sizeof zend, fp) < sizeof zend)
+	File_Seek(fp, -4, SEEK_CUR);
+	if (File_Read(&zend, 1, sizeof zend, fp) < sizeof zend)
 		return true;
 
 	data_size = sizeof zend;
 
 	numlumps = zend.entries;
 
-	fseek(fp, zend.cdiroffset, SEEK_SET);
+	File_Seek(fp, zend.cdiroffset, SEEK_SET);
 	for (i = 0; i < numlumps; i++)
 	{
 		char* fullname;
 		char* trimname;
 		char* dotpos;
 
-		if (fread(&zentry, 1, sizeof(zentry_t), fp) < sizeof(zentry_t))
+		if (File_Read(&zentry, 1, sizeof(zentry_t), fp) < sizeof(zentry_t))
 			return true;
 		if (memcmp(zentry.signature, pat_central, 4))
 			return true;
@@ -2723,7 +3368,7 @@ W_VerifyPK3 (FILE *fp, lumpchecklist_t *checklist, boolean status)
 		if (verified == true)
 		{
 			fullname = malloc(zentry.namelen + 1);
-			if (fgets(fullname, zentry.namelen + 1, fp) != fullname)
+			if (File_GetString(fullname, zentry.namelen + 1, fp) != fullname)
 				return true;
 
 			// Strip away file address and extension for the 8char name.
@@ -2751,30 +3396,30 @@ W_VerifyPK3 (FILE *fp, lumpchecklist_t *checklist, boolean status)
 			free(fullname);
 
 			// skip and ignore comments/extra fields
-			if (fseek(fp, zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
+			if (File_Seek(fp, zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
 				return true;
 		}
 		else
 		{
-			if (fseek(fp, zentry.namelen + zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
+			if (File_Seek(fp, zentry.namelen + zentry.xtralen + zentry.commlen, SEEK_CUR) != 0)
 				return true;
 		}
 
 		data_size +=
 			sizeof zentry + zentry.namelen + zentry.xtralen + zentry.commlen;
 
-		old_position = ftell(fp);
+		old_position = File_Tell(fp);
 
-		if (fseek(fp, zentry.offset, SEEK_SET) != 0)
+		if (File_Seek(fp, zentry.offset, SEEK_SET) != 0)
 			return true;
 
-		if (fread(&zlentry, 1, sizeof(zlentry_t), fp) < sizeof (zlentry_t))
+		if (File_Read(&zlentry, 1, sizeof(zlentry_t), fp) < sizeof (zlentry_t))
 			return true;
 
 		data_size +=
 			sizeof zlentry + zlentry.namelen + zlentry.xtralen + zlentry.compsize;
 
-		fseek(fp, old_position, SEEK_SET);
+		File_Seek(fp, old_position, SEEK_SET);
 	}
 
 	if (data_size < file_size)
@@ -2798,18 +3443,18 @@ W_VerifyPK3 (FILE *fp, lumpchecklist_t *checklist, boolean status)
 // Note: This never opens lumps themselves and therefore doesn't have to
 // deal with compressed lumps.
 static int W_VerifyFile(const char *filename, lumpchecklist_t *checklist,
-	boolean status)
+	fhandletype_t type, boolean status)
 {
-	FILE *handle;
+	void *handle;
 	int goodfile = false;
 
 	if (!checklist)
 		I_Error("No checklist for %s\n", filename);
 	// open wad file
-	if ((handle = W_OpenWadFile(&filename, false)) == NULL)
+	if ((handle = W_OpenWadFile(&filename, type, false)) == NULL)
 		return -1;
 
-	if (stricmp(&filename[strlen(filename) - 4], ".pk3") == 0)
+	if (stricmp(&filename[strlen(filename) - 4], ".pk3") == 0 || stricmp(&filename[strlen(filename) - 4], ".zip") == 0)
 		goodfile = W_VerifyPK3(handle, checklist, status);
 	else
 	{
@@ -2820,7 +3465,7 @@ static int W_VerifyFile(const char *filename, lumpchecklist_t *checklist,
 			goodfile = W_VerifyWAD(handle, checklist, status);
 		}
 	}
-	fclose(handle);
+	File_Close(handle);
 	return goodfile;
 }
 
@@ -2831,13 +3476,14 @@ static int W_VerifyFile(const char *filename, lumpchecklist_t *checklist,
   * be sent.
   *
   * \param filename Filename of the wad to check.
+  * \param type File handle type.
   * \param exit_on_error Whether to exit upon file error.
   * \return 1 if file contains only music/sound lumps, 0 if it contains other
   *         stuff (maps, sprites, dehacked lumps, and so on). -1 if there no
   *         file exists with that filename
   * \author Alam Arias
   */
-int W_VerifyNMUSlumps(const char *filename, boolean exit_on_error)
+int W_VerifyNMUSlumps(const char *filename, fhandletype_t type, boolean exit_on_error)
 {
 	// MIDI, MOD/S3M/IT/XM/OGG/MP3/WAV, WAVE SFX
 	// ENDOOM text and palette lumps
@@ -2928,7 +3574,20 @@ int W_VerifyNMUSlumps(const char *filename, boolean exit_on_error)
 		{NULL, 0},
 	};
 
-	int status = W_VerifyFile(filename, NMUSlist, false);
+	int status = 1;
+
+#if defined(__ANDROID__)
+	if (W_IsAndroidPK3(filename))
+	{
+		void *handle = W_OpenWadFile(&filename, type, false);
+		if (handle == NULL)
+			status = -1;
+		else
+			File_Close(handle);
+	}
+	else
+#endif
+		status = W_VerifyFile(filename, NMUSlist, type, false);
 
 	if (status == -1)
 		W_InitFileError(filename, exit_on_error);
